docs/proposals/201812_thanos-remote-receive.md

```diff
@@ -0,0 +1,166 @@
+
+As the Thanos receivers route the requests to the responsible node, an edge load balancer is configured to randomly distribute load to all Thanos receivers available.
+
+Time-series hashes are calculated from the entire label set of that time-series. To ensure that potentially large time-series with common labels do not all end up being ingested by the same node, the tenant’s ID should be included in the hash. This will help distribute the load across receivers.
+
+The hash is roughly calculated as follows.
+
+```
+hash(string(tenant_id) + sort(timeseries.labelset).join())
+```
+
+The hashing function used is the same one as used by Prometheus’: [xxHash][xxHash]. Sorting of labels is necessary, in order to ensure that a unique time-series always has the same hash.
+
+### Ingestion hard tenancy
+
+In attempts to build similar systems a common fallacy has been to distribute all load from all tenants of the system onto a single set of ingestion nodes. This makes reasoning and management rather simple, however, has turned out to have stronger downsides than upsides. Companies using Cortex in production and offering it to clients have setup entirely separate clusters for different customers, because their load has caused incidents affecting other clients. In order to allow customers to send large amounts of data at irregular intervals the ingestion infrastructure needs to scale without impacting durability. Scaling the ingestion infrastructure can cause endpoints to not accept data temporarily, however, the write-ahead-log based replication can cope with this, as it backs off and continues sending its data once the ingestion infrastructure successfully processes those requests again.
+
+Hard tenants in the Thanos receiver are configured in a configuration file, changes to this configuration must be orchestrated by a configuration management tool. When a remote write request is received by a Thanos receiver it goes through the list of configured hard tenants. For each hard tenant there is a separate hash ring respective to their ingestion nodes as described in the "Load distribution" section. A hard tenant also has the number of associated receive nodes belonging to it. A remote write request can be initially received by any receiver node, however, will only be dispatched to receiver nodes that correspond to that hard tenant.
+
+A sample of the configuration of tenants and their respective infrastructure:
+
+```
+tenants:
+- match: tenant-a
+  nodes:
+  - tenant-a-1.metrics.local
+- match: “”
+  nodes:
+  - soft-tenants-1.metrics.local
+```
+
+To start, only exact matches of tenant IDs will used to distribute requests to receive nodes. Should it be necessary, more sophisticated mechanisms can be added later. When a request is received, the specified tenant is tested against the configured tenant ID until an exact match is found. If the specified tenant is the empty string, then any tenant is considered a valid match.
```

Comment:
I have read the last sentence a few different ways :/ 
I think he we should be explicit that if you do not have `hard tenancy` configured for your write request and you land on a hard tenant your request will be sent to a soft tenant hashring.

Reply:
Actually, you specify that only exact matches to start ... so will not be an initial issue but might be worth clarifying.

Reply:
Added "If no hard tenancy is configured, a tenant will automatically land in a soft tenancy hashring."